File: ios/Classes/KnockPlugin.swift
import Flutter
import UIKit
import Combine

struct FlutterErrorWrapper: Error {
    let flutterError: FlutterError
    let message: String

    init(flutterError: FlutterError) {
        self.flutterError = flutterError
        self.message = flutterError.message ?? "Unknown Flutter error"
    }
}

public class KnockPlugin: NSObject, FlutterPlugin, KnockHostApi {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = KnockPlugin()
        KnockHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
        
        registrar.addApplicationDelegate(instance)
    }
    
    private let apnsTokenSubject = PassthroughSubject<Result<Data, Error>, Error>()

    func getFcmToken(completion: @escaping (Result<String, Error>) -> Void) {
        let flutterError = FlutterError(code: "unsupported", message: "getFcmToken is not supported on iOS", details: nil)

        completion(.failure(FlutterErrorWrapper(flutterError: flutterError)))
    }
    
    func getApnsToken(completion: @escaping (Result<String, Error>) -> Void) {
        var cancellable: AnyCancellable!
        cancellable = apnsTokenSubject
            .receive(on: DispatchQueue.main)
            .sink(receiveCompletion: { _ in }, receiveValue: { result in
                do {
                    let deviceToken = try result.get()
                    let token = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
                    completion(.success(token))
                } catch {
                    completion(.failure(error))
                }
                
                cancellable.cancel()
            })
        
        UIApplication.shared.registerForRemoteNotifications()
    }
}

// MARK:- UIApplicationDelegate
extension KnockPlugin: UIApplicationDelegate {
    public func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        apnsTokenSubject.send(Result.success(deviceToken))
    }
    
    public func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
        apnsTokenSubject.send(Result.failure(error))
    }
}

File: android/src/main/kotlin/app/knock/KnockPlugin.kt
package app.knock

import com.google.firebase.messaging.FirebaseMessaging
import io.flutter.embedding.engine.plugins.FlutterPlugin

class KnockPlugin: FlutterPlugin, KnockHostApi {
  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    KnockHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    KnockHostApi.setUp(binding.binaryMessenger, null)
  }

  override fun getFcmToken(callback: (Result<String>) -> Unit) {
    FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
      if (task.isSuccessful) {
        callback(Result.success(task.result))
      } else {
        callback(Result.failure(Throwable(task.exception)))
      }
    }
  }

  override fun getApnsToken(callback: (Result<String>) -> Unit) {
    callback(Result.failure(FlutterError(
      code = "unsupported",
      message = "getApnsToken is not supported on Android")))
  }
}
